app/Http/Controllers/CityOpinionController.php

```diff
@@ -7,27 +7,47 @@
 use App\Http\Requests\CityOpinionRequest;
 use App\Models\CityOpinion;
 use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Gate;
 
 class CityOpinionController extends Controller
 {
     public function store(CityOpinionRequest $request): void
     {
-        $opinion = $request->only(["rating", "content", "city_id"]);
+        $opinion = $request->validate([
+            "rating" => "required|integer|min:1|max:5",
+            "content" => "required|string",
+            "city_id" => "required|integer",
+        ]);
         $opinion["user_id"] = Auth::id();
 
         CityOpinion::query()->create($opinion);
     }
 
     public function update(CityOpinionRequest $request, CityOpinion $cityOpinion): void
     {
-        $opinion = $request->only(["rating", "content", "city_id"]);
-        $opinion["user_id"] = Auth::id();
-
-        $cityOpinion->update($opinion);
+        $opinion = $request->validate([
+            "rating" => "required|integer|min:1|max:5",
+            "content" => "required|string",
+            "city_id" => "required|integer",
+        ]);
+
+        $response = Gate::inspect("update", $cityOpinion);
+
+        if ($response->allowed()) {
+            $cityOpinion->update($opinion);
+        } else {
+            abort(403, $response->message());
+        }
     }
```

Comment:
Maybe move that into policy.

Reply:
Could you specify what should be moved?

Reply:
`Gate::inspect("update", $cityOpinion);` into routes files. U can check that before entering to the controller in router.